nidx/nidx_relation/src/top_unique_n.rs

```diff
@@ -0,0 +1,116 @@
+// Copyright (C) 2021 Bosutech XXI S.L.
+//
+// nucliadb is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at info@nuclia.com.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+//
+
+use std::collections::HashMap;
+
+/// Top unique N
+///
+/// Maintain the top set of unique keys with greatest scores.
+pub struct TopUniqueN<K> {
+    elements: HashMap<K, f32>,
+    top_n: usize,
+    threshold: f32,
+}
+
+impl<K> TopUniqueN<K>
+where
+    K: Eq + std::hash::Hash + std::fmt::Debug,
+{
+    pub fn new(top_n: usize) -> Self {
+        Self {
+            top_n,
+            elements: HashMap::with_capacity(2 * top_n),
+            threshold: f32::NEG_INFINITY,
+        }
+    }
+
+    pub fn insert(&mut self, key: K, score: f32) {
+        if score < self.threshold {
+            return;
+        }
+
+        if self.elements.len() == self.elements.capacity() {
+            let lowest_score = self.truncate_top_n();
+            self.threshold = lowest_score;
+        }
+
+        self.elements
+            .entry(key)
+            .and_modify(|s| {
+                if score > *s {
+                    *s = score
+                }
+            })
+            .or_insert(score);
+    }
+
+    // Truncate the current set of element to N leaving only the top-scoring
+    // elements. Return the smallest score across the top.
+    fn truncate_top_n(&mut self) -> f32 {
+        let mut vec = Vec::from_iter(self.elements.drain());
+        vec.sort_unstable_by(|a, b| a.1.total_cmp(&b.1).reverse());
+        vec.truncate(self.top_n);
+        let lowest_score = vec.last().expect("truncation must never be done without any element").1;
```

Comment:
You could just return `-infinity` in the case where it's empty and avoid this panic. This can happen if `top_k=0` which we use in some cases withouth graph (e.g: we only want faceted results in catalog). This is safe now , but I can see a future in which we add graph searching to something and suddenly panic because of this.

Reply:
Yes, changed